src/client_side_reply.cc

```diff
@@ -308,6 +308,8 @@ clientReplyContext::processExpired()
     }
 
     if (entry) {
+        entry->createMemObject(url, http->log_uri, http->request->method);
+        entry->mem_obj->setUris(url, http->log_uri, http->request->method);
```

Comment:
This repetition and similar code elsewhere in this PR look really bad. We need to find a better way to express what we know. AFAICT, we need to handle three or four distinct cases (from the caller point of view):

1. The caller knows that the StoreEntry object does not have a MemObject. The caller needs to create the MemObject but the caller does not have the trio information yet. This case is related to the old MemStore problem that we are not going to fix in this PR (and that makes it impossible to make the trio members constant).

2. The caller knows that the StoreEntry object does not have a MemObject. The caller needs to create the MemObject and set its trio information.

3. The caller knows that the StoreEntry object has a MemObject. The caller needs to update the MemObject trio if it is unset. I am not sure these cases actually exist.

4. The caller does not know whether the StoreEntry object has a MemObject member. If the StoreEntry does not have MemObject, then the caller needs to create one (with the caller-supplied trio). Otherwise, the caller needs to update the MemObject trio if it is unset.

I suggest using the following APIs to support all of the above cases:

* `MemObject::MemObject(void)`: Old MemObject constructor. No changes here.
* `MemObject::setUris(trio)`: Old method adjusted to do nothing when hasUris() is already true and to assert that hasUris() becomes true otherwise.
* `StoreEntry::ensureMemObject(trio): Creates MemObject if needed and always calls setUri().
* `StoreEntry::createMemObject(trio): Asserts that there is no MemObject and calls ensureMemObject().
* `StoreEntry::createMemObject(void): Asserts that there is no MemObject and creates one.

When/if we no longer need to update the trio, MemObject(void) and setUris() will disappear, with their code merged into MemObject(trio). The createMemObject(void) method will disappear as well as unneeded.

The use cases listed earlier can use the following APIs:

1. createMemObject(void)
2. createMemObject(trio)
3. `assert(mem_obj); mem_obj->setUris(trio)`
4. ensureMemObject(trio)

Since each case gets its own API, it would be easier to find and adjust them later. Again, I hope that case 3 does not actually exist.

If you do not have a better idea, please see whether the above sketch results in a better looking but still correct code.

Reply:
I decided to implement your sketch "as is".